python_modules/dagster/dagster/_core/events/__init__.py

```diff
@@ -1549,6 +1587,40 @@ def __new__(cls, asset_observation: AssetObservation):
         )
 
 
+@whitelist_for_serdes
+class AssetFailedToMaterializeReason(Enum):
+    COMPUTE_FAILED = "COMPUTE_FAILED"
+    UPSTREAM_COMPUTE_FAILED = "UPSTREAM_COMPUTE_FAILED"
```

Comment:
there's also 'run succeeded without materializing' and 'run was canceled without materializing' right? (which as i understand it is treated like a skip and doesn't turn the asset red, to match current behavior)

Reply:
yeah. i was going to add the other events in a second pass once i had gotten the daemon set up for run failures just to break things up a bit. but no reason i can't add the reason types now

Reply:
took a stab at adding the different reasons i could think of. I also added two lists to split the reasons into the ones that are like failure-failures and ones that are when an asset failed to materialize in an expected way (user canceled a run, optional assets). I don't have to merge those lists in, it's more for trying to communicate the difference between the reasons. 

Basically it's a rough attempt to split apart how the status for an asset in a run should affect the "global" status we display for the asset. For the skipped types, i think we would ideally be able to fall back and show the status of the latest run where the status of the asset is not a skipped type. Not sure how we can do that though with having the "skipped" and "failed" reasons all in one event

Reply:
I'm not totally sure about splitting out user terminated vs system/external terminated. my thinking here is that if a user cancels or deletes a run, then the assets that didn't materialize shouldn't be in a failed state. This is the behavior today. But if a run is killed (by some external process or we cancel it because it runs too long) that should get surfaced as a failure. The thing that's unclear to me is if we are able to split out the run canceled events like that? from a given cancel event are we able to see if it was user initiated or not?